### DISCUSSION TITLE
BlueBanquise stack tools

### DISCUSSION THREAD

**Discussion by oxedions** [bluebanquise/bluebanquise#833] [Ideas]
# BlueBanquise stack proposal

Hello ! 😊 

Collections are close to be stabilized. I want to start thinking about tools to make BlueBanquise more "industrialized".
There are many possibilities, like a WebUI, etc, but for now, I want to focus on CLI tools, keeping in mind possible WebUI later.

Here are some thought, I would be very interested by some feedback 🙂 This is purely "thinking" at this point, to identify field needs.

## Needs

General: KISS, use python and plain text when possible, and make it git compatible for versioning. When needed, allow custom plugins.

### Tool to populate inventory

Inventory would be plain text, YAML or/and INI.

Commands like:

* bluebanquise-inventory node add c001 equipment=supermicro-YYYYY,profile=compute
* bluebanquise-inventory node c001 network_interfaces add interface=eth0,ip4=10.10.0.1,mac=XX:XX:XX:XX:XX,network=net-admin
* bluebanquise-inventory node c001 show
* bluebanquise-inventory equipment add supermicro-XXXXX
* bluebanquise-inventory equipment supermicro-XXXXX update type=server
* bluebanquise-inventory equipment supermicro-XXXXX show
* …

In first version, tool would need to handle:

* add/update/remove node (must be at least in an equipment and a profile)
* add/update/remove node bmc
* add/update/remove node network_interfaces
* manage groups
  * create groupe (including equipments, but these handled differently)
  * add/remove nodes from a group
  * manage equipment groups
* manage main global bb variables

In second version, tool would need to handle:

* add/update network topology, including switchs, cabling, etc. Can be done if handling topology at switchs level and not at nodes level, as it would be easy to loop on switch interfaces to find a specific node nic, while the opposite is not true.
* allow custom keys for nodes and groups.
* allow custom actions/hooks when updates are done.

This tool would be a REST http server, and a cli would be provided to interact with it (so can uses curl via scripts, or the cli tool, or manual inventory files edit).

### Tool to generate playbooks

A simple tool to define which nodes (by profiles) install which roles.

### Tool to check inventory integrity

Simple tool to check for basic traps (same name, same ip, bad ip/mac, bad topo connection, no equipment group, etc). This tool would be called at each inventory update.

### Tool to manage power

Powerman is not flexible enough. Provide a very simple **plugin compatible** tool to manage power, using basic strings swapping (ex: %%IPMI.password%% --> ADMIN). Could allow IPMI, RedFish, but also vendors proprietary commands and tools.

### Scheduler

A simple but powerful scheduler based on RabbitMQ and Celery to execute tasks defined by sys admin, like how to deploy/redeploy a node, how to update a node (ansible playbook), how to etc.
Tool would allow spread of load over all management nodes, and use 2 queues: one CPU intensive (threads limited to prevent servers overloading) and one no CPU intensive (huge amount of threads, to store waiting tasks, like "wait for ssh to come up" or things like that).

The tool could be requested by sys admin to know advancement of a task.

```
bluebanquise-task deploy c001
>> output: task ID 111111 submitted
bluebanquise-task status 111111
>> blablabla...
```

**Comment by oxedions** [comment]
A tool to check inventory : https://github.com/santos-lucas/bluebanquise-lss-custom/blob/main/scripts/validate_bb_inventories.py

**Comment by ginomcevoy** [comment]
Hello, I also have some loose ideas to present about the tool to populate inventory:

- I think the idea is good, dealing with multiple nodes with text inventory is cumbersome. The problem I see is that text inventory is cumbersome for a reason, it is because we need some basic unique information about each node before we can treat the node as "part of the flock". So while we can reduce the effort to create a node, I don't think the problem of providing node details is going away easily.

- Let's say for example we have this command (notice I am using "c002" instead of "c001":
      ```
      bluebanquise-inventory node add c002 equipment=supermicro-YYYYY,profile=compute
      ```
   Here we are not saying what are the unique properties of c002 (network_interfaces is the main issue). I am pretty sure that BB assumes that there is at least one network interface, and it is connected to the "main network", e.g. ice1-1 for single iceberg, or iceX-1 for multiple icebergs. If this network is not present in the inventory, any playbook executed on the equipment group where c002 is will fail.

- Here are some ideas about creating a usable inventory:
       - Always add at least add a network in the main network for the node. The tool can detect if we are in single or multiple iceberg mode, and discover the icebergs.
       -  Assuming the inventory has a single iceberg, add a network entry for c002 in the ice1-1 network. This can only work if we have at least one "c001" node, to extract network details from. For example, we could assume that the tool can detect other cXXX in the 10.10.2.x network segment, and produce "10.20.2.2" for c002. This simplistic guess is dependent on parsing the name, and only goes up to .254, so it is a fragile part of the proposal.
       -  If there are multiple icebergs, the request would fail unless a specific iceberg is provided as input parameter. Once the iceberg X is provided, it will can compute the network segment for the node as above, and add the network interface to the iceX-1 network.
       - The command/API would return a path to the modified text inventory file. The user would then be "recommended" to edit the file to check that the guessed node details are sane, also set a MAC address.

**Reply by ginomcevoy** [reply]
Small extra bit: What I wrote above could be a "version 2" of the tool. For a first version, we could enforce that the "bluebanquise-inventory node add" receives a mandatory parameter for the IP address of the main network. Another command could then be used to customize the network_interface entry.

I think it is important that the resulting inventory from "bluebanquise-inventory node add" does not break functionality when executing roles that read the whole inventory, for example the hosts_file role.

**Reply by oxedions** [reply]
Dear @ginomcevoy 

Many thanks for these feedback! 😊 
Yes, I agree, it would be nice to have such thing. In theory, in BB 2.X, roles will not fail if a node does not have any NIC defined. Node will simply be ignored.

I will upload soon a skeleton of a proposal for the tool and update this page. It should cover some of your requirements 🙂

**Comment by lmagdanello** [comment]
Hey, @oxedions,

I found the idea of ​​Bluebanquise to have a REST API layer to work with inventories very interesting. I didn't find if a draft of this idea already exists, but following what you commented in the discussion, I took the liberty of starting a parallel project focused on this idea:

[bb-api](https://github.com/lmagdanello/bb-api)

For now I just created the BaseModel and a single route [(/nodes](https://github.com/lmagdanello/bb-api/blob/main/app/api/endpoints/node.py)), where we can collect all the nodes declared in the inventory and their main information. I'm using a static inventory and reading it with **_os_**, but we can make this dynamic (perhaps searching /etc/bluebanquise? asynchronously?).

Tell me what you think and if you like the idea, because I'm very interested.

I see that a REST API can also be a great way to use other Bluebanquise tools, even integrating with the administration of the cluster itself (maybe the WEBUI as well? 😃).

**Comment by oxedions** [comment]
Dear @lmagdanello 

Apologies for the delay of my answer, and many thanks for this ! :blush: 

The code looks nice. For the record, all my attemps are stored into BB-6 branch: https://github.com/bluebanquise/bluebanquise/tree/BB-6/stack/tools

However I didn't go far, there was too much things to do on the stack main code. I will help you as much as I can with your developments.
